walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,206 @@
+          } else {
+            endOfViableRange = TimeRange.END_OF_DAY + 1;
+            break;
+          }
+          
+      }
+
+      // Add the last viable range, after we break out of while loop
+      if ((endOfViableRange - startOfViableRange) >= durationOfMeeting && (endOfViableRange - startOfViableRange) < TimeRange.WHOLE_DAY.duration()) {
+        TimeRange lastViableRange = TimeRange.fromStartEnd(startOfViableRange, endOfViableRange, false);
+        viableMeetingDurations.add(lastViableRange);
+      }
+
+      return viableMeetingDurations;
+
+  }
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    // Use Collection so that no duplicate elements (i.e. bad meeting times) are added. No optional attendees considered.
+    Collection<TimeRange> badMeetingTimesNoOptional = new HashSet<>();
+    // If include optional attendees
+    Collection<TimeRange> badMeetingTimesWithOptional = new HashSet<>();
+
+    // viableMeetingDurations is the Collection of viable meeting durations to return
+    // no optional attendees included
+    Collection<TimeRange> viableMeetingDurationsNoOptional;   
+    // with optional attendees 
+    Collection<TimeRange> viableMeetingDurationsWithOptional;
+
+
+    // Convert to int, since MeetingRequest has long duration, but TimeRange has int duration
+    int durationOfMeeting = (int)request.getDuration();
+
+    if(durationOfMeeting > TimeRange.WHOLE_DAY.duration()) {
+        // return empty (no options) for too long of a request (handles the noOptionsForTooLongOfARequest() test)
+        return new ArrayList<TimeRange>();
+    }
+
+    int startOfDay = TimeRange.START_OF_DAY;
+    int endOfDay = TimeRange.END_OF_DAY;
+
+
+    /* For each of the events in the Collection of Events, if the event and the meeting have shared attendees, we 
+    need to signify that the meeting cannot be scheduled during the duration of the event. Therefore, we add the 
+    time range of this event to badMeetingTimes.
+    */
+    for (Event event : events) {
+
+        TimeRange timeRangeOfEvent = event.getWhen();
+
+        // eventAndMeetingShareMandatoryAttendee is true if the event and the meeting share at least one mandatory attendee
+        boolean eventAndMeetingShareMandatoryAttendee = doEventAndMeetingShareAttendee(event, request, true);
+
+         // eventAndMeetingShareAnyAttendee is true if the event and the meeting share at least one attendee, either mandatory or optional
+        boolean eventAndMeetingShareAnyAttendee = doEventAndMeetingShareAttendee(event, request, false);
+
+        if (eventAndMeetingShareMandatoryAttendee) {
+            badMeetingTimesNoOptional.add(timeRangeOfEvent);
+        }
+
+        if (eventAndMeetingShareAnyAttendee) {
+            badMeetingTimesWithOptional.add(timeRangeOfEvent);
+        }
+
+    }
+
```

Comment:
Let's merge overlapped events in badMeetingTimes here, so that we don't need to handle it in line 106 to line 124.